sycl/doc/extensions/proposed/sycl_ext_intel_virtual_functions.asciidoc

```diff
@@ -0,0 +1,579 @@
+
+> The odr-use of polymorphic classes and classes with virtual inheritance is
+> allowed. *However, no virtual member functions are allowed to be called in a
+> device function.*
+
+However, there are still some limitations of how virtual member functions can
+be used:
+
+- if an object is constructed in host code, calling a virtual member function
+  for that object in device code has undefined behavior;
+- if an object is constructed in device code on a device `A`, calling a virtual
+  member function for that object in host code, or on another device `B` has
+  undefined behavior;
+
+=== New properties
+
+Due to the indirect nature of virtual member functions, compiler may not be able
+to understand which exact virtual member function is being called (i.e. which
+class it belongs to).  Moreover, when code is distributed over several
+translation units, compiler may not be able to even see definitions of all
+virtual member functions which may be called in a kernel.
+
+Therefore, to provide a mechanism for an implementation to help detect virtual
+member functions, which are going to be used from kernels and enforce necessary
+restrictions on them, new compile-time-constant properties are proposed.
+
+[source,dpcpp]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  struct indirectly_callable_key {
+    template <typename SetId>
+    using value_t = property_value<indirectly_callable_key, SetId>;
+  };
+
+  struct calls_indirectly_key {
+    template <typename First = void, typename... SetIds>
+    using value_t = property_value<calls_indirectly_key, First, SetIds...>;
+  };
+
+  template <typename SetId = void>
+  inline constexpr indirectly_callable_key::value_t<SetId> indirectly_callable;
+
+  template <typename First = void, typename... Rest>
+  inline constexpr calls_indirectly_key::value_t<First, Rest...>
+      calls_indirectly;
+
+  template <>
+  struct is_property_key<indirectly_callable_key> : std::true_type {};
+  template <> struct is_property_key<calls_indirectly_key> : std::true_type {};
+}
+----
+
+Before describing those properties those properties in more detail, a couple of
+new terms are introduced to simplify the extension specification:
+
+Set of virtual member functions:: a group of virtual member functions which are
+defined with `indirectly_callable` property with the same value of the property
+parameter `SetId`. For the simplicity, this will also be further refferred to as
+a _set_, or as a _set of virtual functions_.
+
+Kernel declares a use of a set of virtual member functions:: a kernel is
+considered to be declaring a use of a set of virtual member functions `SetIdA`
+when it is submitted with `calls_indirectly` property with `SetIdA` included
+into the property parameter `SetIds`. If `SetIdA` is not included into the
+property parameter `SetIds`, or if a kernel is submitted without the property,
+then it is *not* considered to be declaring a use of the set of virtual member
```

Comment:
> Good question. That actually may impact caches that we have for programs, because at least for in-memory cache we don't include properties as a key.

I think this is a general issue with kernel properties that is not specific to `calls_indirectly`.  According to [sycl_ext_oneapi_kernel_properties](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc), the same kernel can be submitted multiple times with different properties.  Therefore, I assume that the caching mechanism would have to include the set of kernel properties as part of the cache key.

However, I notice that @Pennycook did *not* allow the same kernel to be submitted multiple times with different properties with the new APIs in [sycl_ext_oneapi_enqueue_functions](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_enqueue_functions.asciidoc).  Instead, a kernel can only have properties if the kernel is defined as a named kernel object.  Thus, the properties become part of the kernel's type.  If this is the new direction, maybe we should consider changing [sycl_ext_oneapi_kernel_properties](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc), so that it has the same restriction as [sycl_ext_oneapi_enqueue_functions](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_enqueue_functions.asciidoc).  This would mean that the kernel properties would *not* need to be part of the cache key.

> The idea was that the call can't be done at runtime, but it is ok to have it statically in a kernel.

I think we want to allow a compiler to emit a diagnostic if it can prove that a call site will always go to a virtual function that is not in the kernel's set of allowed virtual functions.  In most cases, I think the compiler will not be able to prove this, but in some cases it might be possible.

Reply:
> If this is the new direction, maybe we should consider changing [sycl_ext_oneapi_kernel_properties](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc), so that it has the same restriction as [sycl_ext_oneapi_enqueue_functions](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_enqueue_functions.asciidoc).

I think this makes sense.

This new direction was in part informed by some of the launch queries and forward progress stuff that we drafted after the initial version of sycl_ext_oneapi_kernel_properties.  If a kernel can only ever have one set of properties, querying launch limits is much simpler; if the same kernel can be launched with multiple compile-time properties, those properties have to be provided at query time (always painful, sometimes impossible).

I hadn't updated the old specification because there were so many moving parts, and I wanted to wait for the new direction to settle.  If there are no objections to this simplification, I think we should do it.